File: Leetcode/Leetcode_227_基本计算器II.java
package Leetcode;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

public class 基本计算器II_227 {
    public int calculate(String s) {
        int sign = 1;
        List<Integer> stack_num = new ArrayList<>();
        List<Character> stack_ops = new ArrayList<>();
        int index = 0;
        while (index < s.length()){
            if (s.charAt(index) == '+' || s.charAt(index) == ' '){
                index += 1;
            } else if (s.charAt(index) == '-') {
                sign = -sign;
                index += 1;
            }
            else if (s.charAt(index) == '*' || s.charAt(index) == '/'){
                stack_ops.add(s.charAt(index));
                index += 1;
            }
            else{
                int num = 0;
                while (index < s.length() && Character.isDigit(s.charAt(index))){
                    num = 10 * num + Integer.parseInt(String.valueOf(s.charAt(index)));
                    index += 1;
                }
                if (!stack_ops.isEmpty()){
                    if (stack_ops.get(stack_ops.size() - 1) == '*'){
                        stack_num.add(stack_num.remove(stack_num.size() - 1) * num);
                    }
                    if (stack_ops.get(stack_ops.size() - 1) == '/'){
                        stack_num.add(stack_num.remove(stack_num.size() - 1) / num);
                    }
                    stack_ops.remove(stack_ops.size() - 1);
                }
                else{
                    stack_num.add(sign * num);
                    sign = 1;
                }
            }
        }
        return stack_num.stream().mapToInt(Integer::intValue).sum();
    }
 }

File: Leetcode/Leetcode_227. 基本计算器 II.py
# 227. 基本计算器 II
# 给你一个字符串表达式 s ，请你实现一个基本计算器来计算并返回它的值。
# 整数除法仅保留整数部分。
def calculate_first(s: str):
    length = len(s)
    sign = 1
    index = 0
    stack_num = list()
    stack_ops = list()
    while index < length:
        if s[index] == ' ' or s[index] == '+':
            index = index + 1
        elif s[index] == '-':
            sign = -sign
            index = index + 1
        elif s[index] == '*' or s[index] == '/':
            stack_ops.append(s[index])
            index = index + 1
        else:
            num = 0
            while index < length and s[index].isdigit():
                num = num * 10 + ord(s[index]) - ord('0')
                index = index + 1
            if stack_ops:
                if stack_ops[-1] == '*':
                    stack_num.append(stack_num.pop() * num)
                if stack_ops[-1] == '/':
                    stack_num.append(int(stack_num.pop() / num))
                stack_ops.pop()
            else:
                stack_num.append(sign * num)
                sign = 1
    return sum(stack_num)


s = "3+2*2"
calculate_first(s)
s = " 3/2 "
calculate_first(s)
s = " 3+5 / 2 "
calculate_first(s)
